lint-hrx.ts

```diff
@@ -0,0 +1,161 @@
+
+/**
+ * Verifies all HRX files in the project `spec` directory are under the
+ * {@link LINE_THRESHOLD}.
+ *
+ * HRX files are unwrapped and transformed into real files when
+ * {@link reporter.fix} is true.
+ *
+ * Skips HRX files if they include a top-level `options.yml` file annotated with
+ * :todo: or :ignore-for: `lint-hrx`. For example:
+ *
+ * ```hrx
+ * <===> options.yml
+ * # Ignored because <SOME REASON>
+ * :todo:
+ * - lint-hrx
+ * ```
+ */
+async function lintHrxSize(rootPath: string, reporter: GithubActionsReporter) {
+  const rootDir = await fromPath(rootPath);
+  const tooBig = await bigHrxInTree(rootDir, reporter);
+
+  // No Errors.
+  if (tooBig.length === 0) return;
+
+  if (!reporter.fix) {
+    const tooBigRelative = tooBig.map(f => p.relative(process.cwd(), f));
+    for (const f of tooBigRelative) {
+      reporter.reportError(`HRX file exceeds ${LINE_THRESHOLD} lines`, f, true);
+    }
+    return;
+  }
+
+  for (const archivePath of tooBig) {
+    await hrxToRealFiles(archivePath);
+    // Fixes files recursively.
+    await lintHrxSize(archivePath, reporter);
+  }
+}
+
+/**
+ * Transforms one level of a given HRX file into real files.
+ *
+ * Given:
+ * ```
+ * root
+ * |- foo.hrx        (real)
+ *    |- file1       (virtual)
+ *    |- file2       (virtual)
+ *    |- dir1        (virtual)
+ *    |  |- file2    (virtual)
+ *    |  |- dir2     (virtual)
+ *    |     |- dir3  (virtual)
+ *    |- dir4        (virtual)
+ * ```
+ *
+ * Deletes `foo.hrx` and generates:
+ * ```
+ * root
+ * |- foo            (real)
+ *    |- file1       (real)
+ *    |- file2       (real)
+ *    |- dir1.hrx    (real)
+ *    |  |- file2    (virtual)
+ *    |  |- dir2     (virtual)
+ *    |     |- dir3  (virtual)
+ *    |- dir4.hrx    (real)
+ * ```
+ *
+ */
+async function hrxToRealFiles(archivePath: string) {
+  const virtualDir = (await fromPath(archivePath)) as VirtualDirectory;
+  fs.mkdirSync(virtualDir.path, {recursive: true});
+
+  for (const f of await virtualDir.listFiles()) {
+    const contents = await virtualDir.readFile(f);
+    fs.writeFileSync(p.join(virtualDir.path, f), contents);
+  }
+  for (const d of await virtualDir.subdirs()) {
+    const subArchivePath = `${d.path}.hrx`;
+    fs.writeFileSync(subArchivePath, await d.asArchive());
+  }
+  fs.unlinkSync(archivePath);
+}
+
+/**
+ * Navigates a directory tree to verify all HRX files follow the style guide,
+ * such as number of lines in an HRX file.
+ */
+async function bigHrxInTree(
+  dir: SpecDirectory,
+  reporter: GithubActionsReporter
+): Promise<string[]> {
+  if (!(dir instanceof RealDirectory)) return [];
+
+  return [
+    ...(await bigHRX(dir, reporter)),
+    ...(
+      await Promise.all(
+        (await dir.subdirs()).map(d => bigHrxInTree(d, reporter))
+      )
+    ).flat(),
+  ];
+}
+
+/**
+ * Returns the absolute paths to all HRX files in a single {@link dir} that
+ * exceed the {@link LINE_THRESHOLD} into the global set {@link tooBig}.
```

Comment:
I think this "into the global set" line is outdated?

Reply:
it is 😅